### PR TITLE
Switch to using the pystream eventstream

### PR DESCRIPTION
**Pull request by WillB97**
Moves to a python (aiohttp) based eventstream server (https://github.com/WillB97/srcomp-pystream) that doesn't suffer from a memory leak. Corrects the JSON encoding of unicode characters in team names.

I have checked this installs and run mixtape against it for a week without issues.

### CODE REVIEW

File: modules/compbox/templates/nginx.conf.erb [hunk 1]
```diff
@@ -33,17 +33,18 @@ server {
     }
 
     location = /stream {
+        return 307 /stream/;
```
Comment (PeterJCLaw) [inline modules/compbox/templates/nginx.conf.erb]: In the past we've seen hard-to-diagnose issues with redirecting stream connections (see #13 and https://github.com/PeterJCLaw/srcomp-puppet/commit/9ede7dc0e1d08587d2f1fc1afac4bf096a8bb95b). If possible I think we should maintain the existing url.

**Comment by PeterJCLaw** [comment]
Thanks for this contribution. I'm having a look at the pystream implementation now -- has that been reviewed at all, or does that need doing as part of this change?

> Corrects the JSON encoding of unicode characters in team names.

Please could you expand on what this issue is? I don't recall this being an issue and I can't find any bug reports which mention it.

> I have checked this installs and run mixtape against it for a week without issues.

Did that run include deploys and matches or was it just running against a constant state? The events emitted while nothing is scheduled are quite different than those emitted during a competition event.

It would also be useful to understand what testing has been done to ensure compatibility with the existing stream, unless you're planning also to update all the various clients?

File: modules/compbox/templates/nginx.conf.erb (hunk 1 above)
Comment (WillB97) [reply modules/compbox/templates/nginx.conf.erb]: The other changes here direct the stream directly to `/stream/`. I've added the CORS header to match the stream header and tested it works with both the javascript EventStream and python library.

**Comment by WillB97** [comment]
The current stream leaves raw unicode characters in the JSON output, which is out of spec. This converts them to \u escaped values. Javascript and python's JSON implementations seem to cope with this so it's not a major issue but something that may crop up in other implementations.

In testing I ran it against ~2000 matches, deploying scores, changing the schedule and adding delay. For compatibility I verified the output between this and the previous stream for 200 matches across the same as above.

**Comment by PeterJCLaw** [comment]
> The current stream leaves raw unicode characters in the JSON output, which is out of spec.

Do you have a reference for that? Looking at the spec on https://www.json.org it says:

> A _string_ is a sequence of zero or more Unicode characters, wrapped in double quotes, using backslash escapes.

The image also says (in the box between the quotes):

> Any codepoint except " or \ or control characters

My reading of this is that unicode characters can be directly part of the quoted string. If higher valued codepoints needed converting to a `\u....` then it would say that (or mention that the quoted values needed to be ASCII or something).

Converting them to `\u....` is therefore still valid, however definitely not _required_.

**Comment by PeterJCLaw** [comment]
> In testing I ran it against ~2000 matches, deploying scores, changing the schedule and adding delay. For compatibility I verified the output between this and the previous stream for 200 matches across the same as above.

Nice, that sounds like a good round of tests.

I've looked through the pystream implementation, with a focus on the event construction. I've not delved into the websocket stuff as that's not an area I'm familiar with unfortunately. It would be great to have someone who knows websockets review that at some point. I've done some local testing too and it does seem to (mostly) work as a drop-in replacement for the existing stream.

I've raised some issues & PRs against the pystream repo in the course of reviewing it. Many of them are small things, though I think some of them we should aim to address before moving over:
- https://github.com/WillB97/srcomp-pystream/pull/4 fixes handling of bad responses from the upstream, avoids loading the upstream when those errors happen
- https://github.com/WillB97/srcomp-pystream/issues/9 is the only incompatibly that I found, unfortunately it causes the "outside" screens not to update team names when they change
- https://github.com/WillB97/srcomp-pystream/issues/7 I think needs addressing, though I don't know what the right fix is. It's possible that the current code is already correct or mostly correct and I've misunderstood the semantics, though I suspect that there's some cases which will need changes

If possible it would be great to address https://github.com/WillB97/srcomp-pystream/issues/5 too, hopefully it's a small change, though that probably isn't a blocker to moving over.

File: modules/compbox/templates/nginx.conf.erb [hunk 2]
```diff
@@ -46,6 +46,22 @@ server {
         return 404;
     }
 
+    location = /pystream {
+        add_header Access-Control-Allow-Origin *;
+        return 307 /stream/;
```
Comment (PeterJCLaw) [inline modules/compbox/templates/nginx.conf.erb]:
```suggestion
        return 307 /pystream/;
```

File: modules/compbox/templates/compbox-index.html.erb [hunk 3]
```diff
@@ -28,7 +28,7 @@
     <div class="for-machines">
         <ul>
             <li><a href="/comp-api/">Comp API</a></li>
-            <li><a href="/stream">Comp API Stream</a></li>
+            <li><a href="/stream/">Comp API Stream</a></li>
```
Comment (PeterJCLaw) [inline modules/compbox/templates/compbox-index.html.erb]:
```suggestion
            <li><a href="/stream">Comp API Stream (legacy)</a></li>
            <li><a href="/pystream/">Comp API Stream (experimental)</a></li>
```

File: modules/compbox/templates/screens-config.json.erb [hunk 4]
```diff
@@ -1,4 +1,4 @@
 {
     "apiurl": "/comp-api",
-    "streamurl": "/stream"
+    "streamurl": "/stream/"
```
Comment (PeterJCLaw) [inline modules/compbox/templates/screens-config.json.erb]:
```suggestion
    "streamurl": "/stream"
```
though if we want to move to the new one, which I think we do:
```suggestion
    "streamurl": "/pystream/"
```

File: modules/compbox/manifests/init.pp [hunk 5]
```diff
@@ -339,6 +340,35 @@
         source  => 'puppet:///modules/compbox/stream-404.html',
         notify  => Service['nginx']
     }
+    # pystream
+    package { 'srcomp_pystream':
+        ensure   => $vcs_ensure,
+        provider => 'pip3',
+        source   => "git+https://github.com/WillB97/srcomp-pystream.git"
```
Comment (PeterJCLaw) [inline modules/compbox/manifests/init.pp]: nit:
```suggestion
        source   => 'git+https://github.com/WillB97/srcomp-pystream.git'
```
Double quotes are needed for string interpolation, but puppet linters prefer to use single quotes where there's no interpolation.

**Review by PeterJCLaw — COMMENTED** [review]
Nice, thanks for updating this. Couple of things I suspect were just missed when moving back to having the pystream separate, plus I think we can set the screens to use pystream if you like.

**Comment by PeterJCLaw** [comment]
Thanks! Let's see how this goes tomorrow :)

I realised we also needed to put the pystream in its own folder, so I added that.

**Pull request merged**